### ISSUE TITLE
Ability to remove some of the pre-installed plugins

### ISSUE THREAD

**Issue by kmcquade** [odavid/my-bloody-jenkins#128]
Would be nice to not have all of these plugins installed by default. Suggestion: provide environment variable to specify the pre-installed plugins to uninstall. 

Great repo, by the way. Love the work you’re doing.

**Comment by kmcquade** [comment]
@odavid - any idea when you'd be able to address this? :)

**Comment by odavid** [comment]
Hi @kmcquade - I am sorry, but I was (still am), bit busy lately.

First of all thank you! (: 

As for the feature request, I think that deleting plugins is a bit complex, since plugins have dependencies and dependent plugins + the scripts can fail due to bad import of missing plugin that the image support.
I can always add a script that reads an env var that will delete from the list, but:

*  The plugins will come back in the next startup, since they are baked inside the image
* It won't make the image smaller, since they are already baked within it.

I believe that inheriting from the image and deleting / adding plugins is much safer that doing it during startup.

What problem will it solve you?

Best,
Ohad

**Comment by kmcquade** [comment]
> I believe that inheriting from the image and deleting / adding plugins is much safer than doing it during startup.

Yeah, I agree :) any suggested commands for removing with bash without causing any issues (not sure if removing the .hpi will cause any issues)?

> What problem will it solve you?

Currently, there's a bit of plugin bloat. Examples: GitHub, GitLab, and Bitbucket are not all necessary for any single deployment. HipChat is EOL. We also don't use Checkmarx. 

One question - really looking for your opinion. Performing the plugin configs via JCasC seems much easier than writing Groovy code for every plugin. Do you think the Groovy code method presents any clear benefits over JCasC?

**Comment by odavid** [comment]
Hi @kmcquade - 

> Yeah, I agree :) any suggested commands for removing with bash without causing any issues (not sure if removing the .hpi will cause any issues)?

```shell
# In Dockerfile
RUN rm -rf /usr/share/jenkins/ref/plugins/<PLUGIN-ID>.hpi
```

> One question - really looking for your opinion. Performing the plugin configs via JCasC seems much easier than writing Groovy code for every plugin. Do you think the Groovy code method for plugin configs presents any clear benefits over doing the same with JCasC?

* I wrote and used the image before JCasC  was even Beta - I keep the groovy code, for both backward compatibility reasons and simplicity.
* As you might know, the image supports JCasC within its yaml file. - see https://github.com/odavid/my-bloody-jenkins#configuration-as-code-section
* I think JCasC is great, and I think people should use it.   
*  Sometimes there are things that are not yet supported by JCasC - Such as script-security-plugin 
* Since JCasC is based on the Java POJO structure, the configuration can change if the plugin author re-factors the code. I tried to make a "Groovy Configuration Facade", so even when plugins are changed, the YAML Configuration can be kept the same. 
* Sometimes in Groovy Config Handlers, I tried to define "Good Defaults", such as security options. 

So in summary (: - Since the image supports both jCasC and the Groovy Config Handlers and also can be mixed, I think it is a matter of preference.
I use the groovy handlers for most things, since I have tests that detect plugin compatibility issues, and therefore I know the config will be ok if I upgrade the image version. 
But I also try to adopt jCasC bit by bit ( I also contributed some code to jCasC)

Hope this helps,
Ohad

**Comment by kmcquade** [comment]
Ok great tips. I think I have what I need so I’ll close this issue. You should get on Keybase; would be great to occasionally chat DevOps/DevSecOps with you. 

Appreciate your contributions to FOSS. Cheers.
